File: kakao.app/js/ked/edit/draw.js
var _k_ = {extend: function (c,p) {for (var k in p) { if (Object.prototype.hasOwnProperty(p, k)) c[k] = p[k] } function ctor() { this.constructor = c; } ctor.prototype = p.prototype; c.prototype = new ctor(); c.__super__ = p.prototype; return c;}, in: function (a,l) {return (typeof l === 'string' && typeof a === 'string' && a.length ? '' : []).indexOf.call(l,a) >= 0}, max: function () { var m = -Infinity; for (var a of arguments) { if (Array.isArray(a)) {m = _k_.max.apply(_k_.max,[m].concat(a))} else {var n = parseFloat(a); if(!isNaN(n)){m = n > m ? n : m}}}; return m }, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, eql: function (a,b,s) { var i, k, v; s = (s != null ? s : []); if (Object.is(a,b)) { return true }; if (typeof(a) !== typeof(b)) { return false }; if (!(Array.isArray(a)) && !(typeof(a) === 'object')) { return false }; if (Array.isArray(a)) { if (a.length !== b.length) { return false }; var list = _k_.list(a); for (i = 0; i < list.length; i++) { v = list[i]; s.push(i); if (!_k_.eql(v,b[i],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } } else if (_k_.isStr(a)) { return a === b } else { if (!_k_.eql(Object.keys(a),Object.keys(b))) { return false }; for (k in a) { v = a[k]; s.push(k); if (!_k_.eql(v,b[k],s)) { s.splice(0,s.length); return false }; if (_k_.empty(s)) { return false }; s.pop() } }; return true }, isStr: function (o) {return typeof o === 'string' || o instanceof String}, empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}}

var draw

import kxk from "../../kxk.js"
let kseg = kxk.kseg
let kstr = kxk.kstr

import prof from "../util/prof.js"

import color from "../theme/color.js"
import theme from "../theme/theme.js"

import view from "../view/base/view.js"

import belt from "./tool/belt.js"

import mode from "./mode.js"


draw = (function ()
{
    _k_.extend(draw, view)
    function draw (screen, name, features)
    {
        this["draw"] = this["draw"].bind(this)
        draw.__super__.constructor.call(this,screen,name,features)
        this.setColor('bg',theme.editor.bg)
        this.setColor('empty',theme.editor.empty)
        this.setColor('cursor',theme.cursor)
        this.setColor('selection',theme.selection)
        this.setColor('highlight',theme.highlight)
    }

    draw.prototype["draw"] = function ()
    {
        var bg, c, ch, checkColor, ci, clss, fg, firstIndex, firstSegi, headerClass, line, linel, lines, row, si, syntax, view, x, y, _103_17_, _105_17_, _107_15_, _108_15_, _109_15_, _45_58_, _87_45_

        if (this.hidden())
        {
            return
        }
        syntax = this.state.syntax
        view = this.state.s.view
        lines = this.state.s.lines
        if ((this.complete != null))
        {
            lines = this.complete.preDrawLines(lines)
        }
        bg = this.color.bg
        for (var _a_ = row = 0, _b_ = this.cells.rows; (_a_ <= _b_ ? row < this.cells.rows : row > this.cells.rows); (_a_ <= _b_ ? ++row : --row))
        {
            checkColor = false
            headerClass = null
            y = row + view[1]
            if (y >= lines.length)
            {
                break
            }
            line = lines[y]
            linel = kseg.width(line) - view[0]
            c = 0
            firstIndex = kseg.indexAtWidth(line,view[0])
            firstSegi = kseg.segiAtWidth(line,view[0])
            if (firstIndex !== firstSegi)
            {
                c = 1
            }
            x = 0
            while (x < this.cells.cols)
            {
                ci = x + view[0]
                si = kseg.indexAtWidth(line,ci)
                if (si >= line.length)
                {
                    break
                }
                fg = syntax.getColor(ci,y)
                ch = syntax.getChar(ci,y,line[si])
                if (ch === "#")
                {
                    checkColor = true
                }
                else if (_k_.in(ch,'0█'))
                {
                    clss = syntax.getClass(ci,y)
                    if (clss.endsWith('header'))
                    {
                        headerClass = clss
                    }
                }
                x += ((_87_45_=kseg.segWidth(line[si])) != null ? _87_45_ : 1)
                if (x < this.cells.cols)
                {
                    c += this.cells.add(c,row,ch,fg,bg)
                }
                if (clss === 'invert_bg')
                {
                    bg = this.color.bg
                }
            }
            this.drawRowBackground(row,linel)
            if (checkColor)
            {
                this.drawColorPills(line,row,linel)
            }
            if (headerClass)
            {
                this.drawAsciiHeader(line,row,headerClass)
            }
        }
        this.drawTrailingRows()
        this.drawHighlights()
        this.drawSelections()
        ;(this.complete != null ? this.complete.drawCompletion() : undefined)
        this.drawCursors()
        ;(this.complete != null ? this.complete.drawPopup() : undefined)
        ;(this.gutter != null ? this.gutter.draw() : undefined)
        ;(this.mapscr != null ? this.mapscr.draw() : undefined)
        ;(this.scroll != null ? this.scroll.draw() : undefined)
        mode.postDraw(this.state)
        return draw.__super__.draw.call(this)
    }

    draw.prototype["drawRowBackground"] = function (row, linel)
    {
        if (row + view[1] === this.state.mainCursor()[1])
        {
            if (linel > 0)
            {
                this.cells.bg_rect(0,row,linel,row,this.color.cursor.main)
            }
            if (linel < this.cells.cols)
            {
                return this.cells.bg_fill(_k_.max(0,linel),row,-1,row,this.color.cursor.empty)
            }
        }
        else
        {
            if (linel > 0)
            {
                this.cells.bg_rect(0,row,linel,row,this.color.bg)
            }
            return this.cells.bg_fill(_k_.max(0,linel),row,-1,row,this.color.empty)
        }
    }

    draw.prototype["drawTrailingRows"] = function ()
    {
        var row, vl

        vl = this.state.s.lines.length - this.state.s.view[1]
        if (vl >= this.cells.rows)
        {
            return
        }
        for (var _a_ = row = vl, _b_ = this.cells.rows; (_a_ <= _b_ ? row < this.cells.rows : row > this.cells.rows); (_a_ <= _b_ ? ++row : --row))
        {
            this.cells.bg_fill(0,row,-1,row,this.color.empty)
        }
    }

    draw.prototype["drawHighlights"] = function ()
    {
        var bg, bgc, highlight, hlc, ul, ulc, vx, vy, x, y

        bg = this.color.highlight.bg
        ul = this.color.highlight.ul
        if (!this.cells.screen.t.hasFocus)
        {
            bg = color.darken(bg)
        }
        var _a_ = this.state.s.view; vx = _a_[0]; vy = _a_[1]

        var list = _k_.list(this.state.s.highlights)
        for (var _b_ = 0; _b_ < list.length; _b_++)
        {
            highlight = list[_b_]
            y = highlight[1] - vy
            if (y >= this.cells.rows)
            {
                break
            }
            for (var _c_ = x = highlight[0], _d_ = highlight[2]; (_c_ <= _d_ ? x < highlight[2] : x > highlight[2]); (_c_ <= _d_ ? ++x : --x))
            {
                hlc = this.cells.get_char(x - vx,y)
                switch (hlc)
                {
                    case '{':
                    case '[':
                    case '(':
                    case ')':
                    case ']':
                    case '}':
                        ulc = this.color.highlight.bracket_ul
                        bgc = this.color.highlight.bracket
                        break
                    case "'":
                    case '"':
                        ulc = this.color.highlight.string_ul
                        bgc = this.color.highlight.string
                        break
                    default:
                        ulc = ul
                        bgc = bg
                }

                this.cells.set_bg(x - vx,y,bgc)
                this.cells.set_char(x - vx,y,color.ul_rgb(ulc) + '\x1b[4:1m' + hlc + '\x1b[4:0m')
                this.cells.adjustContrastForHighlight(x - vx,y,bgc)
            }
        }
    }

    draw.prototype["drawSelections"] = function ()
    {
        var bg, li, linebg, selection, spanbg, x, xe, xs, y

        spanbg = this.color.selection.span
        linebg = this.color.selection.line
        if (!this.cells.screen.t.hasFocus)
        {
            spanbg = color.darken(spanbg)
            linebg = color.darken(linebg)
        }
        var list = _k_.list(this.state.s.selections)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            selection = list[_a_]
            bg = (belt.isSpanLineRange(this.state.s.lines,selection) ? spanbg : linebg)
            for (var _b_ = li = selection[1], _c_ = selection[3]; (_b_ <= _c_ ? li <= selection[3] : li >= selection[3]); (_b_ <= _c_ ? ++li : --li))
            {
                y = li - this.state.s.view[1]
                if (y >= this.cells.rows)
                {
                    break
                }
                if (li === selection[1])
                {
                    xs = selection[0]
                }
                else
                {
                    xs = 0
                }
                if (li === selection[3])
                {
                    xe = selection[2]
                }
                else
                {
                    xe = kseg.width(this.state.s.lines[li])
                }
                for (var _d_ = x = xs, _e_ = xe; (_d_ <= _e_ ? x < xe : x > xe); (_d_ <= _e_ ? ++x : --x))
                {
                    this.cells.set_bg(x - this.state.s.view[0],y,bg)
                    this.cells.adjustContrastForHighlight(x,y,bg)
                }
            }
        }
    }

    draw.prototype["drawCursors"] = function ()
    {
        var bg, cursor, fg, mainCursor, s, x, y

        s = this.state.s
        mainCursor = this.state.mainCursor()
        fg = theme.cursor.fg
        bg = mode.themeColor(this.state,'cursor.multi',theme.cursor.multi)
        if (!this.cells.screen.t.hasFocus)
        {
            bg = color.darken(bg)
        }
        var list = _k_.list(s.cursors)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            cursor = list[_a_]
            if (_k_.eql(cursor, mainCursor))
            {
                continue
            }
            if (this.isCursorVisible(cursor))
            {
                this.cells.set_fg_bg(cursor[0] - s.view[0],cursor[1] - s.view[1],fg,bg)
            }
        }
        if (this.isCursorVisible(mainCursor))
        {
            fg = theme.cursor.fg
            bg = mode.themeColor(this.state,'cursor.main',theme.cursor.main)
            if (!this.hasFocus())
            {
                bg = color.darken(bg)
            }
            var _b_ = [mainCursor[0] - s.view[0],mainCursor[1] - s.view[1]]; x = _b_[0]; y = _b_[1]

            if (s.cursors.length <= 1)
            {
                if (this.isCursorInEmpty())
                {
                    bg = color.darken(bg,0.5)
                }
                else if (' ' === this.cells.get_char(x,y))
                {
                    bg = color.darken(bg,0.8)
                }
            }
            if (!this.cells.screen.t.hasFocus)
            {
                bg = color.darken(bg)
            }
            return this.cells.set_fg_bg(x,y,fg,bg)
        }
    }

    draw.prototype["drawColorPills"] = function (line, row, linel)
    {
        var cx, dta, idx, rng, rngs

        if (rngs = kstr.colorRanges(kseg.str(line)))
        {
            cx = _k_.max(0,linel) + 1
            var list = _k_.list(rngs)
            for (idx = 0; idx < list.length; idx++)
            {
                rng = list[idx]
                dta = 4
                if (idx === 0)
                {
                    this.cells.set(cx,row,'',rng.color,this.color.empty)
                    cx += 1
                    dta--
                }
                if (idx === rngs.length - 1)
                {
                    dta--
                }
                this.cells.bg_rect(cx,row,cx + dta,row,rng.color)
                cx += dta
                if (idx === rngs.length - 1)
                {
                    this.cells.set(cx,row,'',rng.color,this.color.empty)
                }
            }
        }
    }

    draw.prototype["drawAsciiHeader"] = function (line, row, clss)
    {
        var bg, ch, chunk, chunks, fg, x

        chunks = kseg.chunks(line)
        var list = _k_.list(chunks)
        for (var _a_ = 0; _a_ < list.length; _a_++)
        {
            chunk = list[_a_]
            if (!(_k_.in(chunk.segl[0],'0█')))
            {
                continue
            }
            x = chunk.index - this.state.s.view[0]
            ch = '▎'
            fg = theme.syntax[clss + ' highlight']
            bg = theme.syntax[clss]
            this.cells.set(x,row,ch,fg,bg)
            if (chunk.segl.length <= 1)
            {
                continue
            }
            x += chunk.segl.length - 1
            ch = '▋'
            fg = theme.syntax[clss]
            bg = theme.syntax[clss + ' shadow']
            this.cells.set(x,row,ch,fg,bg)
        }
    }

    return draw
})()

export default draw;

File: kakao.app/love/lua/edit/draw.lua
--[[
0000000    00000000    0000000   000   000
000   000  000   000  000   000  000 0 000
000   000  0000000    000000000  000000000
000   000  000   000  000   000  000   000
0000000    000   000  000   000  00     00
--]]

view = require "view.base.view"


local draw = class("draw", view)
    
        -- if row+@state.s.view[1] == @state∙mainCursor()[1]
        --     if linel > 0
        --         @cells∙bg_rect 0 row linel row @color.cursor.main
        --     if linel < @cells.cols
        --         @cells∙bg_fill math.max(0 linel) row -1 row @color.cursor.empty
        -- else
        --     if linel > 0
        --         @cells∙bg_rect 0 row linel row @color.bg
        --     @cells∙bg_fill math.max(0 linel) row -1 row @color.empty


function draw:init(name, features) 
        view.init(self, name, features)
        
        self:setColor('bg', theme.editor.bg)
        self:setColor('empty', theme.editor.empty)
        self:setColor('cursor', theme.cursor)
        self:setColor('selection', theme.selection)
        self:setColor('highlight', theme.highlight)
        return self
    end

-- 0000000    00000000    0000000   000   000  
-- 000   000  000   000  000   000  000 0 000  
-- 000   000  0000000    000000000  000000000  
-- 000   000  000   000  000   000  000   000  
-- 0000000    000   000  000   000  00     00  


function draw:draw() 
        if self:hidden() then return end
        
        local view = self.state.s.view
        local lines = self.state.s.lines
        -- log "draw" @cells.cols, @cells.rows, @state.s.lines
        
        if self.complete then 
            lines = self.complete:preDrawLines(lines)
        end
        
        for row in iter(1, self.cells.rows) do 
            local y = ((row + view[2]) - 1)
            if (y > lines:len()) then break end
            
            self:drawLine(lines[y], y, row)
        end
        
        -- @drawTrailingRows()        
        self:drawHighlights()
        self:drawSelections()
        
        if self.complete then 
            self.complete:drawCompletion()
        end
        
        self:render()
        
        if self.gutter then self.gutter:draw() end
        -- if @mapscr ➜ @mapscr∙draw()
        if self.scroll then self.scroll:draw() end
        
        self:drawCursors()
        if self.complete then 
            self.complete:drawPopup()
        end
        
        return mode.postDraw(self.state)
    end

-- ███      ███  ███   ███  ████████
-- ███      ███  ████  ███  ███     
-- ███      ███  ███ █ ███  ███████ 
-- ███      ███  ███  ████  ███     
-- ███████  ███  ███   ███  ████████


function draw:drawLine(line, y, row) 
        -- row ?= y-@state.s.view[2]
        local bg = self.color.bg
        local checkColor = false
        local headerClass = null
        
        local syntax = self.state.syntax
        local view = self.state.s.view
        
        local linel = ((kseg.width(line) + 1) - view[1])
        -- log line, noon(dict.keys(@color))
        local c = 1
        -- firstIndex = kseg.indexAtWidth line view[0] # check if leftmost grapheme is 
        -- firstSegi  = kseg.segiAtWidth line view[0]  # cut in half, if yes, start at
        -- c = 1 if firstIndex != firstSegi            # one column to the right 
        local x = 1
        
        -- log "drawLine" y, line
        
        while (x <= self.cells.cols) do 
            local ci = ((x + view[1]) - 1)
            -- si = kseg.indexAtWidth line ci
            local si = ci
            
            -- log "draw #{@name} #{ci} #{si} #{view} #{r3 kseg.str(line)} #{row} #{@cells.cols} #{@cells.rows} #{y} #{lines.length}"
            
            if (si > line:len()) then 
                -- write ◌c "#{@name} break!"
                break
            end
            
            local fg = syntax:getColor(ci, y)
            
            local ch = syntax:getChar(ci, y, line[si])
            
            if (ch == "#") then checkColor = true end
            -- elif ch == '0' or ch == '█'
            --     clss = syntax∙getClass ci y
            --     if clss.endsWith('header') ➜ headerClass = clss
            
            -- cw = kseg.segWidth line[si]
            local cw = 1
            -- cw = max 1 cw # todo: this is to prevent endless loop with zero width characters, should be fixed eventually
            x = x + cw
            
            if (x <= self.cells.cols) then 
                c = c + (self.cells:add(c, row, ch, fg, bg))
            end
            
            -- if clss == 'invert_bg'
            --     bg = @color.bg
        end
        
        -- @drawRowBackground row linel
        
        if checkColor then self:drawColorPills(line, row, linel) end
        if headerClass then 
    return self:drawAsciiHeader(line, row, headerClass)
        end
    end

-- 00000000    0000000   000   000       0000000     0000000    0000000  000   000   0000000   00000000   0000000    
-- 000   000  000   000  000 0 000       000   000  000   000  000       000  000   000        000   000  000   000  
-- 0000000    000   000  000000000       0000000    000000000  000       0000000    000  0000  0000000    000   000  
-- 000   000  000   000  000   000       000   000  000   000  000       000  000   000   000  000   000  000   000  
-- 000   000   0000000   00     00       0000000    000   000   0000000  000   000   0000000   000   000  0000000    

-- drawRowBackground: row linel ->

-- 000000000  00000000    0000000   000  000      000  000   000   0000000   
--    000     000   000  000   000  000  000      000  0000  000  000        
--    000     0000000    000000000  000  000      000  000 0 000  000  0000  
--    000     000   000  000   000  000  000      000  000  0000  000   000  
--    000     000   000  000   000  000  0000000  000  000   000   0000000   


function draw:drawTrailingRows() 
        -- fill empty rows below last line
        local vl = (self.state.s.lines:len() - self.state.s.view[2])
        if (vl >= self.cells.rows) then return end
        
        for row = vl, self.cells.rows-1 do 
            self.cells:bg_fill(0, row, -1, row, self.color.empty)
        end
    end

-- 000   000  000   0000000   000   000  000      000   0000000   000   000  000000000   0000000  
-- 000   000  000  000        000   000  000      000  000        000   000     000     000       
-- 000000000  000  000  0000  000000000  000      000  000  0000  000000000     000     0000000   
-- 000   000  000  000   000  000   000  000      000  000   000  000   000     000          000  
-- 000   000  000   0000000   000   000  0000000  000   0000000   000   000     000     0000000   


function draw:drawHighlights() 
        local bg = self.color.highlight.bg
        local ul = self.color.highlight.ul
        -- bg = color.darken(bg) if not @cells.screen.t.hasFocus
        
        local vx, vy = unpack(self.state.s.view)
        
        for highlight in array.each(self.state.s.highlights) do 
            local y = (highlight[1] - vy)
            
            if (y >= self.cells.rows) then break end
            
            for x = highlight[1], highlight[3]-1 do 
                local hlc = self.cells:get_char((x - vx), y)
                if (hlc == '{') or (hlc == '[') or (hlc == '(') or (hlc == ')') or (hlc == ']') or (hlc == '}') then local ulc = self.color.highlight.bracket_ul ; local bgc = self.color.highlight.bracket
                elseif (hlc == "'") or (hlc == '"') then local ulc = self.color.highlight.string_ul ; local bgc = self.color.highlight.string
                else local ulc = ul ; local bgc = bg
                end
                
                self.cells:set_bg((x - vx), y, bgc)
                self.cells:set_char((x - vx), y, (((color.ul_rgb(ulc) + '\x1b[4:1m') + hlc) + '\x1b[4:0m'))
                self.cells:adjustContrastForHighlight((x - vx), y, bgc)
            end
        end
    end

--  0000000  00000000  000      00000000   0000000  000000000  000   0000000   000   000   0000000  
-- 000       000       000      000       000          000     000  000   000  0000  000  000       
-- 0000000   0000000   000      0000000   000          000     000  000   000  000 0 000  0000000   
--      000  000       000      000       000          000     000  000   000  000  0000       000  
-- 0000000   00000000  0000000  00000000   0000000     000     000   0000000   000   000  0000000   


function draw:drawSelections() 
        local spanbg = self.color.selection.span
        local linebg = self.color.selection.line
        -- log "drawSelections-" @color.selection
        -- log "drawSelections+" spanbg, linebg
        -- if not @cells.screen.t.hasFocus
        --     spanbg = color.darken spanbg
        --     linebg = color.darken linebg
        
        for selection, si in self.state.s.selections:each() do 
            local bg = (function () 
    if belt.isSpanLineRange(self.state.s.lines, selection) then 
    return spanbg else 
    return linebg
                 end
end)()
            
            for li in iter(selection[2], selection[4]) do 
                local y = ((li - self.state.s.view[2]) + 1)
                
                if (y > self.cells.rows) then break end
                
                local xs = 1
                if (li == selection[2]) then 
                    xs = selection[1]
                end
                
                local xe = 0
                if (li == selection[4]) then 
                    xe = selection[3]
                else 
                    xe = (kseg.width(self.state.s.lines[li]) + 2)
                end
                
                for x = xs, xe-1 do 
                    -- log " " si, " " selection, li, y, @cells.rows, xs, xe, x-@state.s.view[1]
                    self.cells:set_bg(((x - self.state.s.view[1]) + 1), y, bg)
                    -- @cells∙adjustContrastForHighlight x-@state.s.view[1] y bg
                end
            end
        end
    end

--  0000000  000   000  00000000    0000000   0000000   00000000    0000000  
-- 000       000   000  000   000  000       000   000  000   000  000       
-- 000       000   000  0000000    0000000   000   000  0000000    0000000   
-- 000       000   000  000   000       000  000   000  000   000       000  
--  0000000   0000000   000   000  0000000    0000000   000   000  0000000   


function draw:drawCursors() 
        local s = self.state.s
        local mc = self.state:mainCursor()
        
        local fg = self.color.cursor.fg
        
        local bg = mode.themeColor(self.state, 'cursor.multi', self.color.cursor.multi)
        -- bg = color.darken(bg) if not @cells.screen.t.hasFocus
        
        local cursors = s.cursors:arr()
        
        for _, cursor in ipairs(cursors) do 
            if not cursor:eql(mc) then 
                if self:isCursorVisible(cursor) then 
                    local x = ((cursor[1] - s.view[1]) + 1)
                    local y = ((cursor[2] - s.view[2]) + 1)
                    self.cells:draw_rounded_multi_cursor(x, y, bg)
                end
            end
        end
        
        if self:isCursorVisible(mc) then 
            fg = self.color.cursor.fg
            
            bg = mode.themeColor(self.state, 'cursor.main', self.color.cursor.main)
            
            if not self:hasFocus() then bg = color.darken(bg) end
            
            local x = ((mc[1] - s.view[1]) + 1)
            local y = ((mc[2] - s.view[2]) + 1)
            
            if (s.cursors:len() <= 1) then 
                if self:isCursorInEmpty() then 
                    bg = color.darken(bg, 0.5)
                elseif (' ' == self.cells:get_char(x, y)) then 
                    bg = color.darken(bg, 0.8)
                end
            end
            
            -- bg = color.darken(bg) if not @cells.screen.t.hasFocus
            
            return self.cells:draw_rounded_cursor(x, y, bg)
        end
    end

--  0000000   0000000   000       0000000   00000000   00000000   000  000      000       0000000  
-- 000       000   000  000      000   000  000   000  000   000  000  000      000      000       
-- 000       000   000  000      000   000  0000000    00000000   000  000      000      0000000   
-- 000       000   000  000      000   000  000   000  000        000  000      000           000  
--  0000000   0000000   0000000   0000000   000   000  000        000  0000000  0000000  0000000   


function draw:drawColorPills(line, row, linel) 
        local rngs = belt.colorRangesInLine(line)
        if (rngs:len() > 0) then 
            local cx = (linel + 2)
            for idx, rng in ipairs(rngs) do 
                local clr = kstr.hexColor(kseg.str(line:slice(rng[1], rng[2])))
                local dta = 4
                if (idx == 1) then 
                    self.cells:set(cx, row, '', clr, self.color.empty)
                    cx = cx + 1
                    dta = dta - 1
                end
                
                if (idx == rngs:len()) then dta = dta - 1 end
                self.cells:bg_rect(cx, row, (cx + dta), row, clr)
                cx = cx + dta
                if (idx == rngs:len()) then self.cells:set(cx, row, '', clr, self.color.empty) end
            end
        end
    end


function draw:drawAsciiHeader(line, row, clss) 
        -- █ █ █ █ █ █ █ █
        -- █ ▉ ▊ ▋ ▌ ▍ ▏ ▕
        
        -- log "▸ #{kseg.str line}"
        
        local chunks = kseg.chunks(line)
        for _, chunk in ipairs(chunks) do 
            if (chunk.segl[1] == '█') then 
                local x = (chunk.index - self.state.s.view[1])
                local ch = '▎'
                local fg = theme.syntax[(clss + ' highlight')]
                local bg = theme.syntax[clss]
                self.cells.set(x, row, ch, fg, bg)
                
                if (#chunk.segl > 1) then 
                    x = x + ((#chunk.segl - 1))
                    ch = '▋'
                    fg = theme.syntax[clss]
                    bg = theme.syntax[(clss + ' shadow')]
                    self.cells.set(x, row, ch, fg, bg)
                end
            end
        end
    end

return draw